packages/web-components/src/menu/menu.ts

```diff
@@ -0,0 +1,411 @@
+  public openOnHover?: boolean = false;
+
+  /**
+   * Determines if the menu should open on right click.
+   * @public
+   */
+  @observable
+  @attr({ attribute: 'open-on-context', mode: 'boolean' })
+  public openOnContext?: boolean = false;
+
+  /**
+   * Determines if the menu should close on scroll.
+   * @public
+   */
+  @observable
+  @attr({ attribute: 'close-on-scroll', mode: 'boolean' })
+  public closeOnScroll?: boolean = false;
+
+  /**
+   * Determines if the menu open state should persis on click of menu item
+   * @public
+   */
+  @observable
+  @attr({ attribute: 'persist-on-item-click', mode: 'boolean' })
+  public persistOnItemClick?: boolean = false;
+
+  /**
+   * Defines whether the menu is open or not.
+   * @public
+   */
+  @observable
+  @attr({ mode: 'boolean' })
+  public open: boolean = false;
+
+  /**
+   * Holds the slotted menu list.
+   * @public
+   */
+  @observable
+  public slottedMenuList: MenuList[] = [];
+
+  /**
+   * Holds the slotted triggers.
+   * @public
+   */
+  @observable
+  public slottedTriggers: HTMLElement[] = [];
+
+  /**
+   * The positioning container of the menu.
+   * @internal
+   */
+  public positioningContainer?: HTMLElement;
+
+  /**
+   * The trigger element of the menu.
+   * @private
+   */
+  private _trigger?: HTMLElement;
+
+  /**
+   * The menu list element of the menu.
+   * @private
+   */
+  private _menuList?: HTMLElement;
+
+  /**
+   * Holds a reference to a function that is used to cleanup resources.
+   * @public
+   */
+  public cleanup?: () => void;
+
+  /**
+   * Called when the element is connected to the DOM.
+   * Sets up the component.
+   * @public
+   */
+  public connectedCallback() {
+    super.connectedCallback();
+    Updates.enqueue(() => this.setComponent());
+  }
+
+  /**
+   * Called when the element is disconnected from the DOM.
+   * Removes event listeners.
+   * @public
+   */
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+    this.cleanup?.();
+    this.removeListeners();
+  }
+
+  /**
+   * Sets the component.
+   * Sets the trigger and menu list elements and adds event listeners.
+   * @public
+   */
+  public setComponent(): void {
+    if (this.$fastController.isConnected && this.slottedMenuList.length && this.slottedTriggers.length) {
+      this._trigger = this.slottedTriggers![0];
+      this._menuList = this.slottedMenuList![0];
+      this._trigger.setAttribute('aria-haspopup', 'true');
+      this._trigger.setAttribute('aria-expanded', `${this.open}`);
+      this.addListeners();
+    }
+  }
+
+  /**
+   * Toggles the open state of the menu.
+   * @public
+   */
+  public toggleMenu = () => {
+    if (this.open) {
+      this.closeMenu();
+    } else {
+      this.openMenu();
+    }
+  };
+
+  /**
+   * Closes the menu.
+   * @public
+   */
+  public closeMenu = () => {
+    this.open = false;
+    if (this.closeOnScroll) {
+      document.removeEventListener('scroll', this.closeMenu);
+    }
```

Comment:
Edge case: if the `close-on-scroll` attribute changes between open and close, the event listener is never removed.